File: backend-java/src/main/java/web/shopping/demo/Service/CartService.java
package web.shopping.demo.Service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.stereotype.Service;

import web.shopping.demo.DTOs.CartItemDto;
import web.shopping.demo.DTOs.CartResponseDto;
import web.shopping.demo.Mapper.CartMapper;
import web.shopping.demo.Models.CartItem;
import web.shopping.demo.Models.Product;

@Service
public class CartService implements CartServiceImpl
{
    private List<CartItem> items = new ArrayList<>();
    private CartMapper cartMapper;

    public CartService(CartMapper cartMapper){
        this.cartMapper = cartMapper;
    }

    public CartResponseDto getCart() {

        var cartItemList = new ArrayList<CartItemDto>();

        for (CartItem cartItem : items)
            cartItemList.add(cartMapper.cartItemToCartItemDto(cartItem));


        var cartResponseDto = new CartResponseDto();
        cartResponseDto.setItems(cartItemList);

        //Rounding value to 2 decimal places
        double rawValue = cartItemList
                            .stream()
                            .mapToDouble(CartItemDto::getLineTotal)
                            .sum();
        double rounded = Math.round(rawValue * 100.0) / 100.0;

        cartResponseDto.setSubtotal(rounded);

        return cartResponseDto;
    }

    public CartResponseDto addItem(Product product, int quantity) {

        var existingItem = items.stream()
                                .filter(i -> product.getId() == i.getProduct().getId())
                                .findFirst()
                                .orElse(null);

        if (existingItem != null)
            existingItem.setQuantity(existingItem.getQuantity() + quantity);
        else
            items.add(new CartItem(product, quantity));

        return getCart();
    }
}

File: backend-dotnet/Services/CartService.cs
using backend_dotnet.DTOs;
using backend_dotnet.Models;

namespace backend_dotnet.Services;

/*
Leaving interface here for demo purposes but should be in
it's own folder for a cleaner layout and for DI purposes.
*/
public interface ICartService
{
    CartResponseDto GetCart();
    CartResponseDto AddItem(Product product, int quantity);
}

public class CartService : ICartService
{
    private readonly List<CartItem> _items = new();

    public CartResponseDto GetCart()
    {
        var itemDtos = _items.Select(item => new CartItemDto
        {
            ProductId = item.Product.Id,
            Title = item.Product.Title,
            Price = item.Product.Price,
            Quantity = item.Quantity,
            LineTotal = item.Product.Price * item.Quantity
        }).ToList();

        return new CartResponseDto
        {
            Items = itemDtos,
            Subtotal = Math.Round(itemDtos.Sum(i => i.LineTotal), 2)
        };
    }

    public CartResponseDto AddItem(Product product, int quantity)
    {
        var existingItem = _items.FirstOrDefault(i => i.Product.Id == product.Id);

        if (existingItem != null)
            existingItem.Quantity += quantity;
        else
            _items.Add(new CartItem { Product = product, Quantity = quantity });

        return GetCart();
    }
}
